packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1159 @@
+  }) : assert(routeInformationProvider == null || routeInformationParser != null),
+       assert(routerDelegate != null),
+       super(key: key);
+
+  /// The route information provider for the router.
+  ///
+  /// The value at the time of first build will be used as the initial route.
+  /// The [Router] listens to this provider and rebuilds with new names when
+  /// it notifies.
+  ///
+  /// This can be null if this router does not rely on the route information
+  /// to build its content. In such case, the [routeInformationParser] can also be
+  /// null.
+  final RouteInformationProvider routeInformationProvider;
+
+  /// The back button dispatcher for the router.
+  ///
+  /// The two common alternatives are the [RootBackButtonDispatcher] for root
+  /// router, or the [ChildBackButtonDispatcher] for other routers.
+  final BackButtonDispatcher backButtonDispatcher;
+
+  /// The route information parser for the router.
+  ///
+  /// When the [Router] gets a new route information from the [routeInformationProvider],
+  /// the [Router] uses this delegate to parse the route information and produce a
+  /// configuration. The configuration will be used by [routerDelegate] and
+  /// eventually rebuilds the [Router] widget.
+  ///
+  /// Since this delegate is the primary consumer of the [routeInformationProvider],
+  /// it must not be null if [routeInformationProvider] is not null.
+  final RouteInformationParser<T> routeInformationParser;
+
+  /// The router delegate for the router.
+  ///
+  /// This delegate consumes the configuration from [routeInformationParser] and
+  /// builds a navigating widget for the [Router].
+  ///
+  /// It is also the primary respondent for the [backButtonDispatcher]. The
+  /// [Router] relies on the [RouterDelegate.popRoute] to handles the back
+  /// button intends.
+  ///
+  /// If the [RouterDelegate.currentConfiguration] returns a non-null object,
+  /// this [Router] will opt for URL updates.
+  final RouterDelegate<T> routerDelegate;
+
+  /// Retrieve the router of exact type.
+  static Router<R> of<R>(BuildContext context) {
+    final _RouterScope scope = context.dependOnInheritedWidgetOfExactType<_RouterScope>();
+    assert(scope != null);
+    return scope.routerState.widget as Router<R>;
+  }
+
+  /// Forces the [Router] to to run the [callback] with the assumption whether
+  /// it is a navigating event.
+  ///
+  /// Normally, calls to this method is not necessary because the [Router] will
+  /// detect whether the callback is a navigating event automatically as soon
+  /// as it is run. Only use this method when you want to force the [Router]
+  /// to recognize or ignore the [callback] as navigation event.
+  ///
+  /// When [router] detect a navigating event, it will report that event to the
+  /// [routeInformationProvider]. This is useful in the web application because
+  /// the web platform relies on these reporting to support browser backward and
+  /// forward buttons.
```

Comment:
"This is useful in the web application because the web platform relies on these reporting to support browser backward and forward buttons." -> "When a Flutter application is running on the Web, navigation events are used to implement the browser's back and forward buttons."